### PR TITLE
[ISSUE-673] add Truncate component #minor

### LINKED ISSUE #673: 🔮 [New Component] Truncate
Truncate will be a utility component that truncates any text that it is wrapped around

https://primer.style/css/components/truncate

### PR DESCRIPTION
**Pull request by alanbsmith**
Describe your changes here.

This PR adds Truncate and TruncateExpandable components. I think I followed the contributing doc closely, but I haven't created a component for Primer before. So please let me know if I messed something up or it's not inline with your code style.

Closes #673 

### Screenshots

**Truncate**
Screen Shot 2020-02-05 at 12 34 14 AM

**TruncateExpandable**

### Merge checklist
- [x] Added or updated TypeScript definitions (`index.d.ts`) if necessary
- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/components/blob/master/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contibuting guidelines for more infomation on how we review PRs.

### CODE REVIEW

File: src/Truncate.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+Truncate.defaultProps = {
+  as: 'div',
+  maxWidth: 125,
+  theme
+}
+
+Truncate.propTypes = {
+  ...TYPOGRAPHY.propTypes,
+  ...COMMON.propTypes,
+  maxWidth: PropTypes.number,
```
Comment (alanbsmith) [inline src/Truncate.js]: styled-system will take a lot of different propTypes for maxWidth, but it looks like it's predominantly using numbers. I can update that, if you'd like it to be more specific. https://styled-system.com/api#layout

File: src/TruncateExpandable.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import styled from 'styled-components'
+import Truncate from './Truncate'
+
+const TruncateExpandable = styled(Truncate)`
+  &:hover {
+    max-width: 10000px !important;
```
Comment (alanbsmith) [inline src/TruncateExpandable.js]: I felt weird about this `!important` but it's in the [truncate styles](https://github.com/primer/css/blob/master/src/truncate/truncate.scss), so I wanted to mirror that.

File: src/TruncateExpandable.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+  }
+`
+
+export default TruncateExpandable
```
Comment (alanbsmith) [inline src/TruncateExpandable.js]: I debated having `expandable` be just a `variant` on Truncate, but I thought there might be occasions where you'd want to use `target` and `expandable` variants at the same time. Opting to extend the component seems to be the pattern you're following.

File: src/Truncate.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import styled from 'styled-components'
+import {variant, maxWidth} from 'styled-system'
+import PropTypes from 'prop-types'
+import {TYPOGRAPHY, COMMON} from './constants'
+import theme from './theme'
+
+const truncateVariants = variant({
+  variants: {
+    target: {
```
Comment (alanbsmith) [inline src/Truncate.js]: I don't love the name `target` for this variant, as it doesn't feel super descriptive. Really, it's an `inline` modifier for when you're nesting truncated text in another element (probably other text). But `target` was used in the [original spec](https://primer.style/css/components/truncate#truncate-target), so I stuck with it.

Comment (BinaryMuse) [reply src/Truncate.js]: @emplums What have we done with things like this in the past? I also am not a huge fan of `target` as a name for this; I would much prefer a boolean `inline` prop e.g

```jsx
  Some text with a{' '}
  <Truncate as="strong" inline title="branch-name-that-is-really-long">
    branch-name-that-is-really-long
  </Truncate>
```

It feels like the `target` thing is an implementation detail of the CSS version of this component and I would be in favor of changing it here.

File: src/TruncateExpandable.js (hunk 3 above)
Comment (BinaryMuse) [reply src/TruncateExpandable.js]: [Similarly](https://github.com/primer/components/pull/685/files#r375403533), I feel like an `expandable` boolean prop on `Truncate` would be more ergonomic here. Neither really feels like a "variant." But I feel less strongly about this one than I do the `inline` one.

Comment (alanbsmith) [reply src/TruncateExpandable.js]: Thanks for the feedback!

I opted for `variant` based on Brent Jackson's recommendation to [opt for variant enum props over boolean props](https://jxnblk.com/blog/defining-component-apis-in-react/). But agreed that it doesn't feel like a `variant` like `large` or `small`.

That said, the styles for these variations don't currently conflict with one another, so maybe they don't need to be separate components (or variants at all).

**Comment by alanbsmith** [comment]
> "I think it is likely a mistake to try to exactly mirror the semantics of Primer CSS in our React components/props, and that we should be open to making our React components as ergonomic as possible."

Agreed, and I appreciate the suggestions / feedback.

File: src/Truncate.js (hunk 4 above)
Comment (alanbsmith) [reply src/Truncate.js]: After looking at other components, the `inline` boolean prop is a pattern in `CircleBadge` and `ProgressBar`. Apologies for not seeing that earlier. The only difference here would be that `inline` would also add `vertical-align: top`. But that's probably fine. I can make an update to follow that pattern.

**Comment by alanbsmith** [comment]
@BinaryMuse, I pushed up an update based on your feedback. 😄 Thanks again!

File: src/Truncate.js (hunk 4 above)
Comment (emplums) [reply src/Truncate.js]: @BinaryMuse @alanbsmith Whenever we run into something that has a certain naming convention/API in Primer CSS that doesn't feel quite right or ergonomic definitely take the liberty to change it! :) That's part of the reason why we won't ever necessarily have a 1:1 mirror between the different language implementations we've got (CSS, React, Rails) - always better to improve the API than stick with something less than ideal for consistency sake + some languages might have features that make a specific API easier than in other languages so it's fine to deviate a bit :)

File: index.d.ts [hunk 5]
```diff
@@ -275,6 +271,14 @@ declare module '@primer/components' {
     label?: string
   }
 
+  export interface TruncateProps extends StyledSystem.MaxWidthProps, TextProps {
```
Comment (emplums) [inline index.d.ts]: This should extend `Typography` props and `BaseProps` versus `TextProps` (the TextProps is referring to the `Text` component)

File: index.d.ts [hunk 6]
```diff
@@ -275,6 +271,14 @@ declare module '@primer/components' {
...
+    expandable: boolean
+    inline: boolean
+    title: string
```
Comment (emplums) [inline index.d.ts]: I believe `inline` and `expandable` should be optional

File: index.d.ts (hunk 5 above)
Comment (alanbsmith) [reply index.d.ts]: Good catch! Thank you!

File: index.d.ts (hunk 6 above)
Comment (alanbsmith) [reply index.d.ts]: Correct 👍

File: src/Truncate.js (hunk 4 above)
Comment (alanbsmith) [reply src/Truncate.js]: Sounds great. 😄 Updated to `inline` as a boolean prop.

**Comment by alanbsmith** [comment]
@emplums updated! Thanks for the review!

**Pull request merged**